plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -1575,19 +1577,24 @@ public TrinoS3StreamingOutputStream(
 
             this.s3 = requireNonNull(s3, "s3 is null");
 
-            this.buffer = new byte[partSize];
+            this.partSize = partSize;
 
             this.bucketName = requireNonNull(bucketName, "bucketName is null");
             this.key = requireNonNull(key, "key is null");
             this.requestCustomizer = requireNonNull(requestCustomizer, "requestCustomizer is null");
             this.uploadIdFactory = requireNonNull(uploadIdFactory, "uploadIdFactory is null");
             this.uploadExecutor = requireNonNull(uploadExecutor, "uploadExecutor is null");
+            this.buffer = new byte[0];
         }
 
         @Override
         public void write(int b)
                 throws IOException
         {
+            if (!initialized) {
+                buffer = new byte[partSize];
```

Comment:
Could we increment first "buffer" capacity gradually (something similar as `ByteArrayBlockBuilder#growCapacity`)? This would help with tiny writes to multiple partitions.

After first buffer is flushed, then we can just keep it at `partSize` capacity

Reply:
Sure can try that. But keep in mind that the minimal size is 5MB, if you still think increasing from it gradually makes sense then sure I will do it

Reply:
Do I understand correctly based on your comment and on https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html that if there is only 1 part than it can be of any size ?

Reply:
@sopel39 please take a look if this is what you had in mind